src/fu-engine.c

```diff
@@ -990,23 +991,53 @@ fu_engine_verify_update(FuEngine *self, const gchar *device_id, GError **error)
 	return TRUE;
 }
 
+static XbNode *
+fu_engine_get_component_by_guid(FuEngine *self, const gchar *guid)
+{
+	XbNode *component;
+	g_autoptr(GError) error_local = NULL;
+#if LIBXMLB_CHECK_VERSION(0, 3, 0)
+	g_auto(XbQueryContext) context = XB_QUERY_CONTEXT_INIT();
+#endif
+
+	/* no components in silo */
+	if (self->query_component_by_guid == NULL)
+		return NULL;
+
+#if LIBXMLB_CHECK_VERSION(0, 3, 0)
+	xb_query_context_set_flags(&context, XB_QUERY_FLAG_USE_INDEXES);
+	xb_value_bindings_bind_str(xb_query_context_get_bindings(&context), 0, guid, NULL);
+	component = xb_silo_query_first_with_context(self->silo,
+						     self->query_component_by_guid,
+						     &context,
+						     &error_local);
+#else
+	if (!xb_query_bind_str(self->query_component_by_guid, 0, guid, &error_local)) {
+		g_warning("failed to bind 0: %s", error_local->message);
+		return NULL;
+	}
+	component =
+	    xb_silo_query_first_full(self->silo, self->query_component_by_guid, &error_local);
+#endif
+	if (component == NULL) {
+		if (!g_error_matches(error_local, G_IO_ERROR, G_IO_ERROR_NOT_FOUND) &&
+		    !g_error_matches(error_local, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT))
+			g_warning("ignoring: %s", error_local->message);
+		return NULL;
+	}
+	return g_object_ref(component);
+}
+
 XbNode *
 fu_engine_get_component_by_guids(FuEngine *self, FuDevice *device)
 {
 	GPtrArray *guids = fu_device_get_guids(device);
-	g_autoptr(GString) xpath = g_string_new(NULL);
-	g_autoptr(XbNode) component = NULL;
 	for (guint i = 0; i < guids->len; i++) {
 		const gchar *guid = g_ptr_array_index(guids, i);
-		xb_string_append_union(xpath,
-				       "components/component[@type='firmware']/"
-				       "provides/firmware[@type='flashed'][text()='%s']/"
-				       "../..",
-				       guid);
+		XbNode *component = fu_engine_get_component_by_guid(self, guid);
+		if (component != NULL)
+			return component;
```

Comment:
How about `break` here instead and make the final function return `component`?

Reply:
(and move that component declaration up out of the for loop)

Reply:
Yes, I guess that's clearer; fixed.